### PR TITLE
feat: Container Orchestration and Triton Container service improvements

### PR DESCRIPTION
**Pull request by pierantoniomerlino**
> **Note**: We are using the Conventional Commits convention for our pull request titles. Please take a look at the [PR title format document](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#submitting-the-changes) for the supported [types](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#type) and [scopes](https://github.com/eclipse/kura/blob/develop/CONTRIBUTING.md#scope).

This PR improves the `ContainerOrchestrationService` and `TritonContainerService` exposing the following options:

- runtime: set the runtime used by the container engine
- devices: set a list of devices to be passed to the container
- backend path: set the backend folder for the Triton server.

### CODE REVIEW

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml [hunk 1]
```diff
@@ -73,6 +73,15 @@
             description="A comma separated list of inference model names that the server will load.">
         </AD>
 
+        <AD id="local.backends.path"
+            name="Local backends path"
+            type="String"
+            cardinality="0"
+            required="false"
+            default=""
+            description="Specify the path on the filesystem where the backends are stored.">
```
Comment (MMaiero) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]:
```suggestion
            description="Specify the filesystem path where the backends are stored.">
```

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml [hunk 2]
```diff
@@ -110,6 +119,24 @@
             default="">
         </AD>
 
+        <AD id="container.runtime"
+            name="Runtime"
+            description="Specify the runtime the Triton container can use."
```
Comment (MMaiero) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]:
```suggestion
            description="Specifies to the Triton container the runtime to use."
```

Would it make sense to update the descriptions with some examples?

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml [hunk 3]
```diff
@@ -110,6 +119,24 @@
...
+            type="String"
+            cardinality="1"
+            required="false"
+            default="">
+        </AD>
+        
+        <AD id="devices"
+            name="Devices"
+            description="Specify the devices the Triton container can use as a comma separated list of device paths."
```
Comment (MMaiero) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]:
```suggestion
            description="Specifies, as a comma separated list of device paths, the devices that the Triton container can use."
```

File: kura/org.eclipse.kura.container.provider/OSGI-INF/metatype/org.eclipse.kura.container.provider.ContainerInstance.xml [hunk 4]
```diff
@@ -123,6 +123,15 @@
             type="String" cardinality="1"
             required="false" default="" />
             
+        <AD id="container.runtime"
+            name="Runtime"
+            description="Specify the runtime the container can use."
```
Comment (mattdibi) [inline kura/org.eclipse.kura.container.provider/OSGI-INF/metatype/org.eclipse.kura.container.provider.ContainerInstance.xml:128]: Specifies the fully qualified name of an alternate OCI-compatible runtime, which is used to run commands specified by the "run" instruction. Example: "nvidia" corresponds to "--runtime=nvidia".

Comment (mattdibi) [inline kura/org.eclipse.kura.container.provider/OSGI-INF/metatype/org.eclipse.kura.container.provider.ContainerInstance.xml]: I would add a little bit more details here... something like:

```suggestion
            description="Specifies the fully qualified name of an alternate OCI-compatible runtime, which is used to run commands specified by the 'run' instruction. Example: 'nvidia' corresponds to '--runtime=nvidia'."
```

Comment (mattdibi) [reply kura/org.eclipse.kura.container.provider/OSGI-INF/metatype/org.eclipse.kura.container.provider.ContainerInstance.xml]: Also: does this work with Podman? Docker expects fully qualified names, Podman seems[ to only accept paths](https://docs.podman.io/en/latest/markdown/podman-farm-build.1.html#runtime-path). (Yay for consistency BTW)

Comment (mattdibi) [reply kura/org.eclipse.kura.container.provider/OSGI-INF/metatype/org.eclipse.kura.container.provider.ContainerInstance.xml]: Nevermind, I was looking at the wrong docs: https://docs.podman.io/en/latest/markdown/podman.1.html#runtime-value

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml [hunk 5]
```diff
@@ -73,6 +73,15 @@
             description="A comma separated list of inference model names that the server will load.">
         </AD>
 
+        <AD id="local.backends.path"
+            name="Local backends path"
+            type="String"
+            cardinality="0"
+            required="false"
+            default=""
+            description="Specify the filesystem path where the backends are stored.">
```
Comment (mattdibi) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]: I think we should point out that this *overrides the backend stored within the container* with the backend found on the host filesystem. This behaviour is quite nuanced. Could be something like:

```suggestion
            description="Specifies the host filesystem path where the backends are stored. If left blank, the backends provided by the Triton container will be used.">
```

File: kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java [hunk 6]
```diff
@@ -237,18 +240,27 @@ private ContainerConfiguration createContainerConfiguration() {
...
+        }
+        builder.setVolumes(Collections.unmodifiableMap(volumes));
 
         List<String> entrypointOverride = new ArrayList<>();
         entrypointOverride.add("tritonserver");
         entrypointOverride.add("--model-repository=" + TRITON_INTERNAL_MODEL_REPO);
         entrypointOverride.add("--model-control-mode=explicit");
+        if (!this.options.getBackendsPath().isEmpty()) {
+            entrypointOverride.add("--backend-directory=" + TRITON_INTERNAL_BACKENDS_FOLDER);
+        }
```
Comment (mattdibi) [inline kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java:263]: I'm not sure you actually need this: you can mount the backends folder directly where the Triton expects it 🤔 This obviously works too... I don't know what would be the best approach.

Comment (mattdibi) [reply kura/org.eclipse.kura.ai.triton.server/src/main/java/org/eclipse/kura/ai/triton/server/TritonServerContainerManager.java:263]: I would say that the best approach is to stay consistent with what we already have (i.e. `TRITON_INTERNAL_MODEL_REPO` approach). Therefore this is ok as-is.

File: kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml [hunk 7]
```diff
@@ -103,13 +112,31 @@
...
+            description="Specify the runtime the Triton container can use. If the container uses the Nvidia Container Runtime, set it to 'nvidia' and leave the GPUs field empty."
+            type="String"
+            cardinality="1"
+            required="false"
+            default="">
+        </AD>
+        
+        <AD id="devices"
+            name="Devices"
+            description="Specifies, as a comma separated list of device paths, the devices that the Triton container can use."
```
Comment (mattdibi) [inline kura/org.eclipse.kura.ai.triton.server/OSGI-INF/metatype/org.eclipse.kura.ai.triton.server.TritonServerContainerService.xml]:
```suggestion
            description="Specifies, as a comma separated list of device paths, the host devices that the Triton container can access."
```

**Review by mattdibi — CHANGES_REQUESTED** [review]
LGTM but I pointed out a couple of places (mainly in the parameter description) where some more details might be needed.

**Review by mattdibi — APPROVED** [review]

**Pull request merged**